ponix4-src/usr/share/man/man8/acpid.8
.TH acpid 8 ""
.\" Portions Copyright (c) 2001 Sun Microsystems
.\" Portions Copyright (c) Tim Hockin (thockin@hockin.org)
.SH NAME
acpid \- Advanced Configuration and Power Interface event daemon
.SH SYNOPSIS
\fBacpid\fP [\fIoptions\fP]

.SH DESCRIPTION
\fBacpid\fP is designed to notify user-space programs of ACPI events.
\fBacpid\fP should be started during the system boot, and will run as a
background process, by default.  It will open an events file
(\fI/proc/acpi/event\fP by default) and attempt to read whole lines which
represent ACPI events.  If the events file does not exist, \fBacpid\fP will
attempt to connect to the Linux kernel via the input layer and netlink.  When an
ACPI event is received from one of these sources, \fBacpid\fP will examine a
list of rules, and execute the rules that match the event. \fBacpid\fP will
ignore all incoming ACPI events if a lock file exists (\fI/var/lock/acpid\fP by
default).
.PP
\fIRules\fP are defined by simple configuration files.  \fBacpid\fP
will look in a configuration directory (\fI/etc/acpi/events\fP by default),
and parse all regular files with names that consist entirely of upper and
lower case letters, digits, underscores, and hyphens (similar to 
run-parts(8)).
.\" that do not begin with a period ('.') or end with a tilde (~).  
Each file must define two things: an \fIevent\fP and an
\fIaction\fP.  Any blank lines, or lines where the first character is a
hash ('#') are ignored.  Extraneous lines are flagged as warnings, but
are not fatal.  Each line has three tokens: the key, a literal equal sign,
and the value.  The key can be up to 63 characters, and is case-insensitive
(but whitespace matters).  The value can be up to 511 characters, and is
case and whitespace sensitive.
.PP
The event value is a regular expression (see regcomp(3)), against which events are matched.
.PP
The action value is a commandline, which will be invoked via \fI/bin/sh\fP
whenever an event matching the rule in question occurs.  The commandline may
include shell-special characters, and they will be preserved.  The only special
characters in an action value are "%" escaped.  The string "%e" will be
replaced by the literal text of the event for which the action was invoked.
This string may contain spaces, so the commandline must take care to quote the "%e" if it wants a single token.  The string "%%" will be replaced by a
literal "%".  All other "%" escapes are reserved, and will cause a rule to
not load.
.PP
This feature allows multiple rules to be defined for the same event (though no
ordering is guaranteed), as well as one rule to be defined for multiple events.
To force \fBacpid\fP to reload the rule configuration, send it a SIGHUP.
.PP
The pseudo-action \fI<drop>\fP causes the event to be dropped
completely and no further processing undertaken; clients connecting
via the UNIX domain socket (see below) will not be notified of the
event. This may be useful on some machines, such as certain laptops which
generate spurious battery events at frequent intervals. The name of
this pseudo-action may be redefined with a commandline option.
.PP
In addition to rule files, \fBacpid\fP also accepts connections on a UNIX
domain socket (\fI/var/run/acpid.socket\fP by default).  Any application may
connect to this socket.  Once connected, \fBacpid\fP will send the text of
all ACPI events to the client.  The client has the responsibility of filtering
for messages about which it cares.  \fBacpid\fP will not close the client
socket except in the case of a SIGHUP or \fBacpid\fP exiting.
.PP
For faster startup, this socket can be passed in as stdin so that \fBacpid\fP
need not create the socket.  In addition, if a socket is passed in as stdin, 
\fBacpid\fP will not daemonize.  It will be run in foreground.  This behavior 
is provided to support systemd(1).
.PP
.B acpid
will log all of its activities, as well as the stdout and stderr of any
actions, to syslog.
.PP
All the default files and directories can be changed with commandline options.
.SH OPTIONS
.TP 12
.BI \-c "\fR, \fP" \-\-confdir " directory"
This option changes the directory in which \fBacpid\fP looks for rule
configuration files.  Default is \fI/etc/acpi/events\fP.
.TP 12
.BI \-C "\fR, \fP" \-\-clientmax " number"
This option changes the maximum number of non-root socket connections which
can be made to the \fBacpid\fP socket.  Default is \fI256\fP.
.TP 12
.BI \-d "\fR, \fP" \-\-debug
This option increases the \fBacpid\fP debug level by one.  If the debug level
is non-zero, \fBacpid\fP will run in the foreground, and will log to
stderr, in addition to the regular syslog.
.TP
.BI \-e "\fR, \fP" \-\-eventfile " filename"
This option changes the event file from which \fBacpid\fP reads events.
Default is \fI/proc/acpi/event\fP.
.TP
.BI \-n "\fR, \fP" \-\-netlink
This option forces \fBacpid\fP to use the Linux kernel input layer and netlink interface for ACPI events.
.TP
.BI \-f "\fR, \fP" \-\-foreground
This option keeps \fBacpid\fP in the foreground by not forking at startup.
.TP
.BI \-l "\fR, \fP" \-\-logevents
This option tells \fBacpid\fP to log information about all events and actions.
.TP
.BI \-L "\fR, \fP" \-\-lockfile " filename"
This option changes the lock file used to stop event processing.
Default is \fI/var/lock/acpid\fP.
.TP
.BI \-g "\fR, \fP" \-\-socketgroup " groupname"
This option changes the group ownership of the UNIX domain socket to which
\fBacpid\fP publishes events.
.TP
.BI \-m "\fR, \fP" \-\-socketmode " mode"
This option changes the permissions of the UNIX domain socket to which
\fBacpid\fP publishes events.  Default is \fI0666\fP.
.TP
.BI \-s "\fR, \fP" \-\-socketfile " filename"
This option changes the name of the UNIX domain socket which \fBacpid\fP opens.
Default is \fI/var/run/acpid.socket\fP.
.TP
.BI \-S "\fR, \fP" \-\-nosocket " filename"
This option tells \fBacpid\fP not to open a UNIX domain socket.  This
overrides the \fI-s\fP option, and negates all other socket options.
.TP
.BI \-p "\fR, \fP" \-\-pidfile " filename"
This option tells \fBacpid\fP to use the specified file as its pidfile.  If
the file exists, it will be removed and over-written.
Default is \fI/var/run/acpid.pid\fP.
.TP
.BI \-r "\fR, \fP" \-\-dropaction " action"
This option defines the pseudo-action which tells \fBacpid\fP to abort
all processing of an event, including client notifications.
Default is \fI<drop>\fP.
.TP
.BI \-t "\fR, \fP" \-\-tpmutefix
This option enables special handling of the mute button for certain
ThinkPad models with mute LEDs that get out of sync with the mute state
when the mute button is held down.  With this option, the mute button
will generate the following events in sync with the number of presses
(and, by extension, the state of the LED):
.IP
.br
button/mute MUTE (key pressed) K
.br
button/mute MUTE (key released) K
.TP
.BI \-v "\fR, \fP" \-\-version
Print version information and exit.
.TP
.BI \-h "\fR, \fP" \-\-help
Show help and exit.
.SH EXAMPLE
This example will shut down your system if you press the power button.
.PP
Create a file named /etc/acpi/events/power that contains the following:
.IP
.br
event=button/power
.br
action=/etc/acpi/power.sh "%e"
.PP
Then create a file named /etc/acpi/power.sh that contains the following:
.IP
/sbin/shutdown \-h now "Power button pressed"
.PP
Now, when \fBacpid\fP is running, a press of the power button will cause the
rule in /etc/acpi/events/power to trigger the script in /etc/acpi/power.sh.
The script will then shut down the system.
.SH TROUBLESHOOTING
\fBacpid\fP is a simple program that runs scripts in response to ACPI
events from the kernel.  When there's trouble, the problem is rarely
with \fBacpid\fP itself.  The following are some suggestions for
finding the most common sources of ACPI-related problems.
.PP
When troubleshooting \fBacpid\fP, it is important to be aware that other parts
of a system might be handling ACPI events.
\fBsystemd\fP(1) is capable of handling the power switch and various other
events that are commonly handled by \fBacpid\fP.  See the description of
HandlePowerKey in \fBlogind.conf\fP(5) for more.  Some window managers also
take over \fBacpid\fP's normal handling of the power button and other events.
.PP
kacpimon(8) can be used to verify that the expected ACPI events are
coming in.  See the man page for kacpimon(8) for the proper procedure.
If the events aren't
coming in, you've probably got a kernel driver issue.
.PP
If the expected events are coming in, then you'll need
to check and see if your window manager is responsible for
handling these events.  Some are, some aren't.  (E.g. in Ubuntu
14.04 (Unity/GNOME), there are settings for the laptop lid in the
System Settings > Power > "When the lid is closed" fields.)
If your window manager is responsible for handling the problematic
event, and you've got it configured properly, then you may have a
window manager issue.
.PP
Lastly, take a look in \fI/etc/acpi/events\fP (see above).  Is there
a configuration file in there for the event
in question (e.g. /etc/acpi/events/lidbtn for laptop lid open/close
events)?  Is it properly connected to a
script (e.g. /etc/acpi/lid.sh)?  Is that script working?  It's not
unusual for an \fBacpid\fP script to check and see if there is a window
manager running, then do nothing if there is.  This means it is up
to the window manager to handle this event.
.SH DEPENDENCIES
\fBacpid\fP should work on any linux kernel released since 2003.
.SH FILES
.PD 0
.B /proc/acpi/event
.br
.B /dev/input/event*
.br
.B /etc/acpi/
.br
.B /var/run/acpid.socket
.br
.B /var/run/acpid.pid
.br
.B /var/lock/acpid
.br
.PD
.SH BUGS
There are no known bugs.  To file bug reports, see \fBPROJECT WEBSITE\fP 
below.
.SH SEE ALSO
regcomp(3), sh(1), socket(2), connect(2), init(1), systemd(1), 
acpi_listen(8), kacpimon(8)
.SH PROJECT WEBSITE
http://sourceforge.net/projects/acpid2/
.SH AUTHORS
Ted Felix <ted@tedfelix.com>
.br
Tim Hockin <thockin@hockin.org>
.br
Andrew Henroid



ponix4-const/history.txt
2011-02-03 19:45 - 2.00.575-beta - added graphical libs to usb-ip package
2011-02-04 14:45 - 2.00.576-beta -
2011-02-07 18:10 - 2.00.577-beta - build MTRR (vesa not working)
2011-02-07 18:19 - 2.00.578-beta -
2011-02-10 20:02 - 2.00.579-beta - libsane 1.0.21 with avision backend from git
2011-02-25 18:09 - 2.00.580-beta - edit libsane package
2011-02-25 18:16 - 2.00.581-beta -
2011-02-27 19:50 - 2.00.582-beta - usb-server package first build. included like package
2011-02-28 21:00 - 2.00.583-beta - final release
2011-02-28 21:07 - 2.00.584 - final release
2011-02-28 21:21 - 2.00.585 - see previos message
2011-02-28 21:45 - 2.1.001 - new version format
2011-02-28 22:51 - 2.1.1 - new version
2011-02-28 23:38 - 2.1.2 - sane package
2011-03-01 20:50 - 2.1.3 - new package sane 1.0.22
2011-03-01 23:05 - 2.1.4 -
2011-03-31 22:06 - 2.1.5 - lzma for package and usb2com converter
2011-04-18 16:44 - 2.1.6 - rxvt.pkg
2011-04-18 18:33 - 2.1.7 - rxvt.pkg ver2
2011-04-18 18:41 - 2.1.8 - rxvt ver3
2011-05-06 15:25 - 2.1.9 - repair freerdp server connect
2011-06-29 21:06 - 2.1.10 - old ver
2011-06-29 23:52 - 2.1.11 - intel video driver 2.15.0, freerdp from git, package rdesktop 1.7
2011-06-30 00:06 - 2.1.12 -
2011-06-30 00:34 - 2.1.13 -
2011-06-30 13:27 - 2.1.14 -
2011-06-30 14:20 - 2.1.15 -
2011-06-30 15:14 - 2.1.16 -
2011-06-30 15:40 - 2.1.17 -
2011-06-30 16:04 - 2.1.18 -
2011-07-01 00:11 - 2.1.19 -
2011-07-01 01:15 - 2.1.20 -
2011-07-05 20:51 - 2.1.21-standalone - rename to standalone version
2011-07-06 14:16 - 2.1.22 - add intel driver 2.15.0 and last version libdrm
2011-07-06 18:37 - 2.1.23 - rdesktop+keymaps
2011-07-07 16:05 - 2.1.24 - add keymaps to the freerdp17 package
2011-07-07 19:16 - 2.1.25 - error in the rdesktop package
2011-07-07 19:50 - 2.1.26 - too
2011-07-07 20:10 - 2.1.27 - vmview auto start
2011-07-07 20:16 - 2.1.28 - too
2011-07-07 21:08 - 2.1.29 - edit vmopenclient
2011-07-08 00:11 - 2.1.30 - edit vmopenclient
2011-07-12 21:18 - 2.1.31 - etoken
2011-07-13 18:28 - 2.1.32 - repair sane package
2011-07-13 19:02 - 2.1.33 - add smart-card support to rdesktop17 package
2011-07-21 13:38 - 2.1.34 - last version of realtek ethernet driver r8101
2011-07-21 14:52 - 2.1.35 - add firmware fo rtl8105e
2011-07-21 19:18 - 2.1.36 - new version of realtek r8169 driver (15)
2011-07-21 20:55 - 2.1.37 - without r8101.ko
2011-07-21 21:27 - 2.1.38 - r8169 from kernek, r8101 delete
2011-07-22 16:45 - 2.1.39 - add citrix client package
2011-07-22 16:51 - 2.1.40 -
2011-07-22 19:03 - 2.1.41 -
2011-07-22 19:59 - 2.1.42 -
2011-07-22 20:07 - 2.1.43 -
2011-07-22 20:40 - 2.1.44 -
2011-07-23 14:11 - 2.1.45-standalone - trying wifi build with odinary /etc/init.d/network (no ethtool)
2011-07-26 13:16 - 2.1.46 -
2011-07-26 15:20 - 2.1.47 -
2011-07-26 16:22 - 2.1.48 -
2011-07-27 14:30 - 2.1.49 -
2011-07-27 15:43 - 2.1.50 -
2011-07-27 18:05 - 2.1.51 -
2011-08-01 13:59 - 2.1.52 -
2011-08-01 15:36 - 2.1.53 -
2011-08-01 16:25 - 2.1.54 -
2011-08-01 18:18 - 2.1.55 -
2011-08-01 20:09 - 2.1.56 -
2011-08-01 20:10 - 2.1.57 - renew xorg server and libXfont
2011-08-01 20:43 - 2.1.58 -
2011-08-01 21:40 - 2.1.59 -
2011-08-01 22:32 - 2.1.60 -
2011-08-01 23:24 - 2.1.61 -
2011-08-01 23:37 - 2.1.62 -
2011-08-02 13:53 - 2.1.63 -
2011-08-02 15:07 - 2.1.64 -
2011-08-02 19:31 - 2.1.65 -
2011-08-02 20:30 - 2.1.66 -
2011-08-02 21:16 - 2.1.67 -
2011-08-02 22:29 - 2.1.68 -
2011-08-02 22:37 - 2.1.69 -
2011-08-03 16:08 - 2.1.70 -
2011-08-03 19:12 - 2.1.71 -
2011-08-03 19:47 - 2.1.72 - ICA 11 packages working!
2011-08-03 20:34 - 2.1.73 -
2011-08-03 20:43 - 2.1.74 -
2011-08-10 22:32 - 2.1.75 - add no_tls and no_nla to freerdp parameters
2011-08-11 00:30 - 2.1.76-standalone -
2011-08-11 18:35 - 2.1.77 - repair bug with freerdp_no_nla freerdp_no_tls
2011-08-11 21:45 - 2.1.78 - repair bug with Xorg (absent libfreetype)
2011-08-12 15:15 - 2.1.79 - add relatime parameter to udev script
2011-08-16 15:07 - 2.1.80 -
2011-08-16 17:26 - 2.1.81 -
2011-08-16 17:51 - 2.1.82 -
2011-08-16 21:43 - 2.1.83 - add gstreamer for ica11 package
2011-08-16 22:31 - 2.1.84 - add gstreamer and plugins for ica11 package
2011-08-18 13:57 - 2.1.85 - add gstream for speedscreen function for citrix client
2011-08-18 14:47 - 2.1.86 - add oss support to the alsa drivers
2011-08-18 15:11 - 2.1.87 -
2011-08-18 16:08 - 2.1.88 - rebuild libasound
2011-08-18 18:01 - 2.1.89 - error, soundcore must be from the kernel
2011-08-18 18:55 - 2.1.90 -
2011-09-12 18:06 - 2.1.91 - correct error with rdesktop17 package add pcsclite library
2011-09-12 22:31 - 2.1.92 - add icewm
2011-09-12 23:00 - 2.1.93 - add graphical libs to icewm
2011-09-12 23:11 - 2.1.94 - add locales to icewm
2011-09-12 23:37 - 2.1.95 -
2011-09-12 23:45 - 2.1.96 -
2011-09-13 00:07 - 2.1.97 -
2011-09-13 00:22 - 2.1.98 -
2011-09-13 13:32 - 2.1.99 -
2011-09-13 14:17 - 2.1.100 - 1
2011-09-13 19:30 - 2.1.101 -
2011-09-13 20:01 - 2.1.102 -
2011-09-13 20:40 - 2.1.103 -
2011-09-13 23:41 - 2.1.104 -
2011-09-14 15:00 - 2.1.105 -
2011-09-14 15:37 - 2.1.106 -
2011-09-16 15:08 - 2.1.107 -
2011-09-26 21:38 - 2.1.108 - in network_script added SERVER_IP=$serverid. hack to use dhcp server as tftp when no siaddr specified (current udhcpc has a bug and doesn't set tftp option)
2011-09-27 20:49 - 2.1.109 - last version of nvidia xorg driver
2011-09-28 14:48 - 2.1.110 - add --no-nla parameter for freerdp client
2011-10-03 20:00 - 2.1.111 - new version of atheros ethernet driver atl1e (AR81Family-linux-1.0.1.14) instead atl1c and atl1e (now removed)
2011-10-03 20:24 - 2.1.112-standalone - current update
2011-10-03 20:52 - 2.1.113 -
2011-10-03 21:22 - 2.1.114 -
2011-10-03 22:10 - 2.1.115 - udev-debug
2011-10-03 22:56 - 2.1.116 - udev debug
2011-10-03 23:23 - 2.1.117 - normal udev
2011-10-03 23:36 - 2.1.118 - udev debug
2011-10-03 23:42 - 2.1.119 - udev debug
2011-10-04 14:48 - 2.1.120 - corrected bug with network_script
2011-10-04 18:28 - 2.1.121 - add atl1c
2011-10-04 18:41 - 2.1.122 -
2011-10-04 19:32 - 2.1.123 - rollback to 1.0.0.9 atl1e driver

ponix4-src/source/acpid-2.0.23/acpid.c
/*
 *  acpid.c - ACPI daemon
 *
 *  Portions Copyright (C) 2000 Andrew Henroid
 *  Portions Copyright (C) 2001 Sun Microsystems
 *  Portions Copyright (C) 2004 Tim Hockin (thockin@hockin.org)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <signal.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <getopt.h>
#include <dirent.h>

#include "acpid.h"
#include "log.h"
#include "event.h"
#include "connection_list.h"
#include "proc.h"
#include "sock.h"
#include "input_layer.h"
#include "inotify_handler.h"
#include "netlink.h"

static int handle_cmdline(int *argc, char ***argv);
static void close_fds(void);
static int daemonize(void);
static void open_log(void);
static int std2null(void);
static int create_pidfile(void);
static void clean_exit(int sig);
static void reload_conf(int sig);

/* global debug level */
int acpid_debug;

/* do we log event info? */
int logevents;

const char *progname;
static const char *confdir = ACPID_CONFDIR;
static const char *lockfile = ACPID_LOCKFILE;
static int nosocket;
static int foreground;
static const char *pidfile = ACPID_PIDFILE;
static int netlink;
const char *dropaction = DROP_ACTION;
int tpmutefix = 0;

int
main(int argc, char **argv)
{
	/* learn who we really are */
	progname = (const char *)strrchr(argv[0], '/');
	progname = progname ? (progname + 1) : argv[0];

	/* handle the commandline  */
	handle_cmdline(&argc, &argv);

	/* close any extra file descriptors */
	close_fds();

	/* open the log */
	open_log();
	
	if (!netlink) {
		/* open the acpi event file in the proc fs */
		/* if the open fails, try netlink */
		if (open_proc())
			netlink = 1;
	}

	if (netlink) {
		/* open the input layer */
		open_input();

		/* watch for new input layer devices */
		open_inotify();

		/* open netlink */
		open_netlink();
	}

	/* open our socket */
	if (!nosocket) {
		open_sock();
	}

	/* if we're running in the background, and we're not being started */
	/* by systemd */
	if (!foreground  &&  !is_socket(STDIN_FILENO)) {
		if (daemonize() < 0)
			exit(EXIT_FAILURE);
	}

	/* redirect standard files to /dev/null */
	if (std2null() < 0) {
		exit(EXIT_FAILURE);
	}
	
	acpid_log(LOG_INFO, "starting up with %s",
		netlink ? "netlink and the input layer" : "proc fs");

	/* trap key signals */
	signal(SIGHUP, reload_conf);
	signal(SIGINT, clean_exit);
	signal(SIGQUIT, clean_exit);
	signal(SIGTERM, clean_exit);
	signal(SIGPIPE, SIG_IGN);

	/* read in our configuration */
	if (acpid_read_conf(confdir)) {
		exit(EXIT_FAILURE);
	}

	/* create our pidfile */
	if (!foreground && create_pidfile() < 0) {
		exit(EXIT_FAILURE);
	}

	acpid_log(LOG_INFO, "waiting for events: event logging is %s",
	    logevents ? "on" : "off");

	/* main loop */
	while (1) {
		fd_set readfds;
		int nready;
		int i;
		struct connection *p;

		/* it's going to get clobbered, so use a copy */
		readfds = *get_fdset();

		/* wait on data */
		nready = TEMP_FAILURE_RETRY(select(get_highestfd() + 1, &readfds, NULL, NULL, NULL));

		if (nready < 0) {
			acpid_log(LOG_ERR, "select(): %s", strerror(errno));
			continue;
		}

		/* house keeping */
		acpid_close_dead_clients();

		/* for each connection */
		for (i = 0; i <= get_number_of_connections(); ++i) {
			int fd;

			p = get_connection(i);

			/* if this connection is invalid, bail */
			if (!p)
				break;

			/* get the file descriptor */
			fd = p->fd;

			/* if this file descriptor has data waiting */
			if (FD_ISSET(fd, &readfds)) {
				/* delegate to this connection's process function */
				p->process(fd);
			}
		}
	}

	clean_exit_with_status(EXIT_SUCCESS);

	return 0;
}

/*
 * Parse command line arguments
 */
static int
handle_cmdline(int *argc, char ***argv)
{
	struct option opts[] = {
		{"confdir", 1, 0, 'c'},
		{"clientmax", 1, 0, 'C'},
		{"debug", 0, 0, 'd'},
		{"eventfile", 1, 0, 'e'},
		{"foreground", 0, 0, 'f'},
		{"logevents", 0, 0, 'l'},
		{"socketgroup", 1, 0, 'g'},
		{"socketmode", 1, 0, 'm'},
		{"socketfile", 1, 0, 's'},
		{"nosocket", 1, 0, 'S'},
		{"pidfile", 1, 0, 'p'},
		{"lockfile", 1, 0, 'L'},
		{"netlink", 0, 0, 'n'},
		{"dropaction", 1, 0, 'r'},
		{"tpmutefix", 0, 0, 't'},
		{"version", 0, 0, 'v'},
		{"help", 0, 0, 'h'},
		{NULL, 0, 0, 0},
	};
	const char *opts_help[] = {
		"Set the configuration directory.",	/* confdir */
		"Set the limit on non-root socket connections.",/* clientmax */
		"Increase debugging level (implies -f).",/* debug */
		"Use the specified file for events.",	/* eventfile */
		"Run in the foreground.",		/* foreground */
		"Log all event activity.",		/* logevents */
		"Set the group on the socket file.",	/* socketgroup */
		"Set the permissions on the socket file.",/* socketmode */
		"Use the specified socket file.",	/* socketfile */
		"Do not listen on a UNIX socket (overrides -s).",/* nosocket */
		"Use the specified PID file.",		/* pidfile */
		"Use the specified lockfile to stop processing.", /* lockfile */
		"Force netlink/input layer mode. (overrides -e)", /* netlink */
		"Define the pseudo-action to drop an event.", /* dropaction */
		"Fixup for ThinkPad mute-repeat behaviour.", /* tpmutefix */
		"Print version information.",		/* version */
		"Print this message.",			/* help */
	};
	struct option *opt;
	const char **hlp;
	int max, size;

	for (;;) {
		int i;
		i = getopt_long(*argc, *argv,
		    "c:C:de:flg:m:s:Sp:L:nr:tvh", opts, NULL);
		if (i == -1) {
			break;
		}
		switch (i) {
		case 'c':
			confdir = optarg;
			break;
		case 'C':
			clientmax = strtol(optarg, NULL, 0);
			break;
		case 'd':
			foreground = 1;
			acpid_debug++;
            log_debug_to_stderr = 1;
			break;
		case 'e':
			eventfile = optarg;
			break;
		case 'f':
			foreground = 1;
			break;
		case 'l':
			logevents = 1;
			break;
		case 'g':
			socketgroup = optarg;
			break;
		case 'm':
			socketmode = strtol(optarg, NULL, 8);
			break;
		case 's':
			socketfile = optarg;
			break;
		case 'S':
			nosocket = 1;
			break;
		case 'p':
			pidfile = optarg;
			break;
		case 'L':
			lockfile = optarg;
			break;
		case 'n':
			netlink = 1;
			break;
		case 'r':
			dropaction = optarg;
			break;
		case 't':
			tpmutefix = 1;
			break;
		case 'v':
			printf(PACKAGE "-" VERSION "\n");
			exit(EXIT_SUCCESS);
		case 'h':
		default:
			fprintf(stderr, "Usage: %s [OPTIONS]\n", progname);
			max = 0;
			for (opt = opts; opt->name; opt++) {
				size = strlen(opt->name);
				if (size > max)
					max = size;
			}
			for (opt = opts, hlp = opts_help;
			     opt->name;
			     opt++, hlp++) {
				fprintf(stderr, "  -%c, --%s",
					opt->val, opt->name);
				size = strlen(opt->name);
				for (; size < max; size++)
					fprintf(stderr, " ");
				fprintf(stderr, "  %s\n", *hlp);
			}
			exit(EXIT_FAILURE);
			break;
		}
	}

	*argc -= optind;
	*argv += optind;

	return 0;
}

static void
close_fds(void)
{
    struct dirent *dent;
    DIR *dirp;
    char *endp;
    long fd;

    if ((dirp = opendir("/proc/self/fd")) != NULL) {
        while ((dent = readdir(dirp)) != NULL) {
            fd = strtol(dent->d_name, &endp, 10);
            if (dent->d_name != endp && *endp == '\0' &&
                fd >= 3 && fd != dirfd(dirp)) {
                close((int) fd);
            }
        }
        closedir(dirp);
    }
}
static int
daemonize(void)
{
	pid_t pid, sid;

	/* fork off the parent process */
	pid = fork();
	if (pid < 0) {
		acpid_log(LOG_ERR, "fork: %s", strerror(errno));
		return -1;
	}
	/* if we got a good PID, then we can exit the parent process */
	if (pid > 0) {
		exit(EXIT_SUCCESS);
	}

	/* at this point we are executing as the child process */

	/* change the umask to something predictable instead of inheriting */
	/* whatever from the parent */
	umask(0);

	/* create a new SID for the child process and */
	/* detach the process from the parent (normally a shell) */
	sid = setsid();
	if (sid < 0) {
		acpid_log(LOG_ERR, "setsid: %s", strerror(errno));
		return -1;
	}

    /* Change the current working directory.  This prevents the current
       directory from being locked; hence not being able to remove it. */
	if (chdir("/") < 0) {
		acpid_log(LOG_ERR, "chdir(\"/\"): %s", strerror(errno));
		return -1;
	}

	return 0;
}

static void
open_log(void)
{
	int log_opts;

	/* open the syslog */
	log_opts = LOG_CONS|LOG_NDELAY;
	if (acpid_debug) {
		log_opts |= LOG_PERROR;
	}
	openlog(PACKAGE, log_opts, LOG_DAEMON);
}

static int
std2null(void)
{
	int nullfd;

	/* open /dev/null */
	nullfd = open("/dev/null", O_RDWR);
	if (nullfd < 0) {
		acpid_log(LOG_ERR, "can't open /dev/null: %s", strerror(errno));
		return -1;
	}

	/* set up stdin, stdout, stderr to /dev/null */

	/* don't redirect stdin if we're being sent a socket by systemd */
	if (!is_socket(STDIN_FILENO)  && 
			dup2(nullfd, STDIN_FILENO) != STDIN_FILENO) {
		acpid_log(LOG_ERR, "dup2() stdin: %s", strerror(errno));
		close(nullfd);
		return -1;
	}
	if (!acpid_debug && dup2(nullfd, STDOUT_FILENO) != STDOUT_FILENO) {
		acpid_log(LOG_ERR, "dup2() stdout: %s", strerror(errno));
		close(nullfd);
		return -1;
	}
	if (!acpid_debug && dup2(nullfd, STDERR_FILENO) != STDERR_FILENO) {
		acpid_log(LOG_ERR, "dup2() stderr: %s", strerror(errno));
		close(nullfd);
		return -1;
	}

	close(nullfd);

	return 0;
}

static int
create_pidfile(void)
{
	int fd;

	/* JIC */
	unlink(pidfile);

	/* open the pidfile */
	fd = open(pidfile, O_WRONLY|O_CREAT|O_EXCL, 0644);
	if (fd >= 0) {
		FILE *f;

		/* write our pid to it */
		f = fdopen(fd, "w");
		if (f != NULL) {
			fprintf(f, "%d\n", getpid());
			fclose(f);
			/* leave the fd open */
			return 0;
		}
		close(fd);
	}

	/* something went wrong */
	acpid_log(LOG_ERR, "can't create pidfile %s: %s",
		    pidfile, strerror(errno));
	return -1;
}

void
clean_exit_with_status(int status)
{
	acpid_cleanup_rules(1);
	delete_all_connections();
	acpid_log(LOG_NOTICE, "exiting");
	unlink(pidfile);
	exit(status);
}

static void
clean_exit(int sig __attribute__((unused)))
{
	clean_exit_with_status(EXIT_SUCCESS);
}

static void
reload_conf(int sig __attribute__((unused)))
{
	acpid_log(LOG_NOTICE, "reloading configuration");
	acpid_cleanup_rules(0);
	acpid_read_conf(confdir);
}

int
locked()
{
	struct stat trash;

	/* check for existence of a lockfile */
	return (stat(lockfile, &trash) == 0);
}

